rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,161 @@
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike an oauth authorization code, static credentials pose a much greater risk of
+exposure, and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name # name must match the name of the plugin.
+spec:
+  last_rotated: "2023-05-09T16:50:52.497874Z"
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a plugin requires static credentials, there is a 1-to-1 mapping between `Plugin`
+objects and `PluginStaticCredentials` objects. If a `Plugin` is using oauth, however,
+it will not need a `PluginStaticCredentials` and none will be created. The object will
+inherit the same name as its associated plugin.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`PluginNeedsStaticCredentials` with a single boolean `Enabled` flag. If this flag is set,
+this will indicate to Teleport that the plugin must have a corresponding
+`PluginNeedsStaticCredentials` object in order to start.
+
+### Access control
+
+`PluginStaticCredentials` can be written to and deleted by any user with RBAC permissions
+to do so. For these operations, the object is straightforward and reflects the typical
+way that Teleport resources are handled.
+
+However, the object will be only readable by the `RoleAdmin` role (with one exception).
+This is because plugins run on the auth server, so it's the only role that needs to be
+able to actually read the static credentials.
+
+The one exception here is Teleport Assist, which runs in the proxy. The proxy will be
+able to read plugin credentials with the name of `openai-default`, which is the name of
+the Assist plugin.
+
+### Lifecycle of the `PluginStaticCredentials` object
+
+#### Creation
+
+A `PluginStaticCredentials` object can be created after a `Plugin` object has been
```

Comment:
Is this just a convenience thing? We could get rid of this half-initialized state by requiring the static credentials to be created before the plugin itself. In the Web UI workflow, this would be handled by code. In any manual onboarding workflow, we would need to guide the user.

Of course then we have to also think about what happens when we delete the secret before the plugin: do we fail the operation, or do we handle this in the plugin manager as mentioned below:

> <...> stop [the plugin instance] if the credentials have been removed for a given plugin.

So I'm not sure whether such upfront validation is worth it.

Reply:
Ack, that sounds good to me. Let's do that, it sounds like less of a dance to get a plugin working.